### ISSUE TITLE
[New] calculate delay value per country in parameters forecast

### ISSUE THREAD

**Issue by Inglezos** [lisphilar/covid19-sir#513]
## Summary of this new feature
In parameters forecast calculate dynamically delay instead of fixed value (recovery period) for each country.

## Proposed solution
Calculate delay maybe by comparing the dates per country when the index was changed rapidly or critical measures were imposed to the dates of the phases or the dates when parameters changed rapidly.
Like applying change point analysis but in parameters-index plane instead of S-R.
Averaging of these change points duration then could indicate such delay period.

**Comment by lisphilar** [comment]
With #280, we implemented `Scenario.fit_predict()` to predict the ODE parameter values of the future phases (< about 17 days = average recovery period). However, this should not be a fixed value.
As @rebeccadavidsson commented, correlation of OxCGRT stringency index and Rt values (as representative values) is expected to be considered.

Dear @rebeccadavidsson, could you please explain the details?

FYI:
"Delay period" is specified in `Scenario._fit_create_data` of the current version 2.14.0.

We will (may) add a method to estimate "delay period" and use the result in `.fit()` as default.

**Comment by rebeccadavidsson** [comment]
I think as @inglezos commented, the delay period could be calculated by analyzing rapid change of the index but it's also possible that no rapid change can be seen for various countries. As I demonstrated in [#280](https://github.com/lisphilar/covid19-sir/issues/280), it would be expected that the strength correlation between the ODE parameters and the Stringency Index would increase along with time, but only after a delay of n number of days. However, the strength of correlation seems to increase already after one day and carries on to increase, until about 20-25 days. 
So I agree with change point detection analysis for each country, for each ODE parameter and averaging this out. The point at which the correlation coefficiënt changes direction could then possibly indicate the end of the effect of policy measures.

![image](https://user-images.githubusercontent.com/11540873/104049855-ecf87a80-51e5-11eb-94e4-ce05db8492f9.png)

**Comment by lisphilar** [comment]
Does the gray span in your notebook mean the effective range of delay period?
https://github.com/rebeccadavidsson/SIR_LSTM/blob/main/corr_oxf.ipynb

If I understand your ideas correctly, we will apply change point analysis on correlation coefficient of Rt and Stringency index for delay period candidates and select best value of delay period (when correlation coefficient changes from high to low).

**Comment by rebeccadavidsson** [comment]
The grey area indicates the window where changes in Rt are **expected**, so the changes are expected from at least 10 days after increase of stringency index. Then, using change point analysis on the correlation coefficient of Rt and Stringency Index (SI) it could be indicated until which day such effects of strengthening the SI can be seen. If the correlation coefficient goes down - or actually up, considering a negative correlation, so let's go with a 'change in correlation coefficient' - this could be an indication of the ending effect from SI strengthening. 

And indeed, change point analysis of the correlation coefficient of Rt and SI could indicate the delay period, so the time it takes before effects can be seen on Rt. However, I think the changes in Rt are too small to choose this as an indicator for the delay period for every country, but I will look into that in more detail tomorrow.

**Comment by lisphilar** [comment]
Thank you for your detailed explanation. I agree with your idea to use the correlation coefficient of Rt and SI for finding delay period. The way to select the exact delay period is to be determined, but this can implemented to CovsirPhy as an experimental method.
Could you try to create a pull request later when you have time? @Inglezos and I will support.

Dear @Inglezos, do you have any ideas regarding this issue?

**Comment by rebeccadavidsson** [comment]
Sounds good. I will, and I will keep you up to date 👍

**Comment by Inglezos** [comment]
Yes, I think that this implementation consists a good first approach for the subject.

**Comment by rebeccadavidsson** [comment]
Briefly looked into [Change Point Analysis and gradient analysis ](https://github.com/rebeccadavidsson/SIR_LSTM/blob/main/ChangePointAnalysis.ipynb) but in early stages it already became clear that the changes are too small to select a specific day for each country. Maybe this is a good approach for countries where correlation with Stringency Index is very strong, but that's nearly never the case. Stays a hard problem to estimate the delay.

**Comment by Inglezos** [comment]
What if we perform the change point analysis in the Daily Confirmed <-> Stringency Index plane instead? I suppose the measures affect directly the new confirmed cases (either they stabilize or decrease).

**Comment by Inglezos** [comment]
I developed a first draft code to test that idea:
https://colab.research.google.com/gist/Inglezos/27a38bcbf752ed95699ee5940ba8c148/untitled0.ipynb

It seems to give normal delays around 15-25 days. We can tweak the ruptures parameters or make further enhancements.

**Comment by rebeccadavidsson** [comment]
I agree that it is better to look at Confirmed - Stringency Index relationship instead of Rt. Tweaking the parameters such as min_size looks promising and could help towards the problem that the delay days can range to +25 days. I am also not ruling out the possibility that the large numbers are correct for countries that don't have the correct tools for data collection and registration of cases.

**Comment by Inglezos** [comment]
I think that unfortunately these large delays might be representative for some countries indeed, considering the data collection or the more lenient measures. I would expect the delays in general to vary somewhere between 2-4 weeks. Also one important factor is the people obedience to these measures. So actually, the fact that the government announces these measures doesn't mean necessarily that they will be applied immediately and thus, the confirmed I suppose describe the actual situation.

**Comment by lisphilar** [comment]
Thank you for your notebooks.
I agree with the idea to calculate delay period using confirmed data and OxCGRT indicators because the OxCGRT indicators are more strongly linked with rho parameter (confirmation) than sigma and so on.
Indicators are listed on https://lisphilar.github.io/covid19-sir/usage_quick.html#Short-term-prediction-of-parameter-values

We may need to collect data regading actions to improve medical service, including new medicines and new hospitals, but this is the next issue for improved forecasting.

Dear @rebeccadavidsson, please remove `df[col] = df.groupby("Country")[col].shift(0 - delay)` from In[411] of "(Experimental): Relationship of OxCGRT index and parameter values" section to avoid shifting OxCGRT data in preprocessing. This may change the output.

**Comment by rebeccadavidsson** [comment]
Thank you for your input @lisphilar. I looked into the output after removing this shift but results are still too vague.

Also, @Inglezos,[ I added one line](https://colab.research.google.com/drive/1NDXluLvzQF6UcvnfJRFD_5wRdgehpmhy) removing the outliers of very long periods, since these very long periods for more than 60 days (approximately) probably won't contribute much to estimating the delay period.

**Comment by Inglezos** [comment]
You mean the outliers right before `df["periods"].mean()`? Yes that would be more appropriate.

**Comment by Inglezos** [comment]
So how shall we proceed with this issue?

**Comment by rebeccadavidsson** [comment]
Good question... After excluding long periods (+30 days), using the technique from your code, average delay for all countries (excluding 8 countries) was 16.3 days. On other words, using changepoint analysis, average period was 16.3 days, averaged out for all countries. However there are still so many factors that influence the length of these periods. There is also no clear related literature that uses this idea to look at the delay of effects. Because of uncertainty and the number of factors that influence the data, I think this is not the best approach, but indeed serves as a working first solution. Ideas for further implementation could be to look at separate policy measure factors (school closing, gatherings restrictions etc.) or to combine this idea with SR analysis.

![delay_days](https://user-images.githubusercontent.com/11540873/104435498-c2fbdb00-558c-11eb-876d-d43b6ed73cf7.png)

**Comment by lisphilar** [comment]
How about implement the method to calculate deley period with some arguments (`indicator="Stringency Index"`, `target="Confirmed"`) and `Warning("This is an experimental method. arguments and solutions may be changed in the future.)`.
Because CovsirPhy package is a tool of anaysis, it is required to provide tools that users can controll with aruguments and try/test many ideas. Our project may publish papers, but I'm not planning and GitHub Pages is the best tool for documentation for rapid development at this time.

**Comment by Inglezos** [comment]
Yes I totally agree, the functionality shall be controllable by such important arguments.
One question, the days of the forecast are equal to this period? Because I think the `fit_predict()` method currently does not accept `forecast_days` for example right?

**Comment by lisphilar** [comment]
> the days of the forecast are equal to this period? Because I think the fit_predict() method currently does not accept forecast_days for example right?

Yes `forecast_days` is equal to `delay` period at this time because OxCGRT data is shifted by `delay` (=`self.jhu_data.recovery_period` at version 2.14.0) in line 1020 and 1021 of scenario.py.

When delay period is 17 days and today is 13Jan2021,
- In `.fit()`, linear regression model learns the relationship of indicators on 01Jan2020 and parameters on 18Jan2020,..., that of indicators on 27Dec2020 and parameters on 13Jan2021.
- In `.predict()`, linear regression model predict parameters on 14Jan2021 using indicators on 28Dec2020,..., predict parameters on 30Jan2021 using indicators on 13Jan2021.

**Comment by Inglezos** [comment]
Yes I understand. So we can never forecast for longer that period days right? Is this a natural barrier that always has to restrict us or we can if we further enhance the predictor somehow?

**Comment by lisphilar** [comment]
Yes, I think as you said. It could be overcome, but long forecasting means prediction of OxCGRT indicators.

`Scenario` class should have methods to add future phases with user-defined OxCGRT scores (issue #477) and `OxCGRTData` class should have methods to convert user-friendly arguments to OxCGRT scores (new issue if necessary), but I do not think prediction of OxCGRT scores with automated solutions is a good approach for COVID-19 analysis. This is because policy measures are actions of ourselves.

**Comment by Inglezos** [comment]
Yes I don't think that it would be very useful to predict OxCGRT indicators. The whole point is to analyze actual government measures to see how they will affect the cases.

**Comment by rebeccadavidsson** [comment]
> How about implement the method to calculate deley period with some arguments (`indicator="Stringency Index"`, `target="Confirmed"`) and `Warning("This is an experimental method. Arguments and solutions may be changed in the future.)`?
> Because CovsirPhy package is a tool of anaysis, it is required to provide tools that users can controll with aruguments and try/test many ideas. Our project may publish papers, but I'm not planning and GitHub Pages is the best tool for documentation for rapid development at this time.

I also think this is a good idea. Then also maybe add the arguments ```minimum days of delay``` and ```maximum days of delay```, where the default is around 7 and 30 respectively.

**Comment by lisphilar** [comment]
Thank you. Conding steps will be
1. add `.Scenario.estimate_delay(oxcgrt_data, indicator="Stringency Index", target="Confirmed", value_range=(7, 30)` (names can be changed) to estimate delay period with change point detection of indicator-target correlation. `self.record_df[target]` will be used to get confirmed data internally.
2. add `delay=None` argument to `.fit()` and `._fit_create_data()`. When `None`, `delay` will be estimated with `.estimate_delay()`.
3. add `**kwargs` to `.fit()` and  `.estimate_delay()` to show figures related to change point detection using `analysis.datahandler.DataHandler.line_plot()` method. (`DataHandler` class is a parent class of `Scenario`.)

@rebeccadavidsson , do you have time to try (some of) them?

**Comment by rebeccadavidsson** [comment]
Yes, great. I think the function ```Scenario.estimate_delay()``` should also return a list or dataframe with the estimated periods and/or the corresponding dates. And that there is a possibility to have "Rt" as target-argument.

And yes, I have time to try these functions 👍 .

**Comment by lisphilar** [comment]
Thank you! The details will be discussed in pull requests.

**Comment by lisphilar** [comment]
Dear @rebeccadavidsson, how is it coming along? :-)

**Comment by rebeccadavidsson** [comment]
Hi @lisphilar, will look at it tomorrow, sorry for the wait!

**Comment by Inglezos** [comment]
I think it is more accurate to use (by default) the "Confirmed" as target rather than the "Infected", because "Infected" is an indirect category which depends on our processing of recovered data. "Confirmed" would be more representative I guess.

Also change default upper range to 45 days (one and a half month)? For example I know for Greece for sure the 2nd wave measures affected and decreased the cases after ~1.5 month.

**Comment by rebeccadavidsson** [comment]
I agree with using Confirmed as default. And very interesting about Greece, where did you deduce that from?

**Comment by Inglezos** [comment]
I am from Greece and living in Greece all this time and from daily reports I am watching everyday the news about the cases. In my opinion the measures seemed to take actual effect and thus reduce drastically the measures in late December. While the measures were first announced in early November. So 35 days limit as default is not accurate since there is at least one country with probably same or more days delay.

**Comment by lisphilar** [comment]
How can we proceed this discussion with data?
I agree with the idea to use Confirmed, but the daily new case data (Confirmed.diff()) will be used.

**Comment by lisphilar** [comment]
Close for new release.

**Comment by Inglezos** [comment]
>but the daily new case data (Confirmed.diff()) will be used.
>
Why is the `diff()` used? Until the last post, I thought the infected were being used.
What I felt more correct was the confirmed to be used instead and the range to be extended to (7, 45).
